public/modules/custom/grants_handler/src/ApplicationHandler.php

```diff
@@ -557,6 +556,75 @@ public static function createApplicationNumber(WebformSubmission $submission): s
     return 'GRANTS-' . $appParam . '-' . $typeCode . '-' . sprintf('%08d', $serial);
   }
 
+  /**
+   * Generate next available application number for the submission.
+   *
+   * @param \Drupal\webform\Entity\WebformSubmission $submission
+   *   Webform data.
+   *
+   * @return string
+   *   Generated number.
+   */
+  public static function getAvailableApplicationNumber(WebformSubmission &$submission): string {
+    $appParam = self::getAppEnv();
+    $serial = $submission->serial();
+    $webform_id = $submission->getWebform()->id();
+    $lastSerialKey = $webform_id . '_' . $appParam;
+    $kvService = \Drupal::service('keyvalue.database');
+    $kvStorage = $kvService->get('application_numbers');
+    $savedSerial = $kvStorage->get($lastSerialKey);
+
+    if (!empty($submission->getData())) {
+      return self::createApplicationNumber($submission);
+    }
+
+    if ($savedSerial && $savedSerial > $serial) {
+      $serial = $savedSerial;
+    }
+
+    $atvService = \Drupal::service('helfi_atv.atv_service');
+
+    $applicationType = $submission->getWebform()
+      ->getThirdPartySetting('grants_metadata', 'applicationType');
+
+    $typeCode = self::getApplicationTypes()[$applicationType]['code'] ?? '';
+
+    $check = FALSE;
+    while ($check === FALSE) {
+      $applicationNumber = self::getApplicationNumberInEnvFormat($appParam, $typeCode, $serial);
+      $applNumberIsAvailable = $atvService->checkDocumentExistsByTransactionId($applicationNumber);
```

Comment:
![image](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/892920/9c11694a-9309-4dc1-a92f-87416c0f8555)